x-pack/test/saved_object_api_integration/common/lib/saved_object_test_utils.ts

```diff
@@ -0,0 +1,302 @@
+      }
+      expect(object).not.to.have.property('error');
+      expect(object.updated_at).to.match(/^[\d-]{10}T[\d:\.]{12}Z$/);
+      // don't test attributes, version, or references
+    }
+  },
+  /**
+   * Additional assertions that we use in `bulk_create` and `create` to ensure that
+   * newly-created (or overwritten) objects don't have unexpected properties
+   */
+  successCreated: async (es: any, spaceId: string, type: string, id: string) => {
+    const isNamespaceUndefined =
+      spaceId === SPACES.DEFAULT.spaceId || isNamespaceAgnostic(type) || isMultiNamespace(type);
+    const expectedSpacePrefix = isNamespaceUndefined ? '' : `${spaceId}:`;
+    const savedObject = await es.get({
+      id: `${expectedSpacePrefix}${type}:${id}`,
+      index: '.kibana',
+    });
+    const { namespace: actualNamespace, namespaces: actualNamespaces } = savedObject._source;
+    if (isNamespaceUndefined) {
+      expect(actualNamespace).to.eql(undefined);
+    } else {
+      expect(actualNamespace).to.eql(spaceId);
+    }
+    if (isMultiNamespace(type)) {
+      if (id === CASES.MULTI_NAMESPACE_DEFAULT_AND_SPACE_1.id) {
+        expect(actualNamespaces).to.eql([DEFAULT_SPACE_ID, SPACE_1_ID]);
+      } else if (id === CASES.MULTI_NAMESPACE_ONLY_SPACE_1.id) {
+        expect(actualNamespaces).to.eql([SPACE_1_ID]);
+      } else if (id === CASES.MULTI_NAMESPACE_ONLY_SPACE_2.id) {
+        expect(actualNamespaces).to.eql([SPACE_2_ID]);
+      } else {
+        // newly created in this space
+        expect(actualNamespaces).to.eql([spaceId]);
+      }
+    }
+    return savedObject;
+  },
+};
+
+/**
+ * Get test scenarios for each type of suite.
+ * @param modifier Use this to generate additional permutations of test scenarios.
+ *  For instance, a modifier of ['foo', 'bar'] would return
+ *  a `securityAndSpaces` of: [
+ *    { spaceId: DEFAULT_SPACE_ID, users: {...}, modifier: 'foo' },
+ *    { spaceId: DEFAULT_SPACE_ID, users: {...}, modifier: 'bar' },
+ *    { spaceId: SPACE_1_ID, users: {...}, modifier: 'foo' },
+ *    { spaceId: SPACE_1_ID, users: {...}, modifier: 'bar' },
+ *  ]
+ */
+export const getTestScenarios = <T>(modifier?: T[]) => {
+  const commonUsers = {
+    noAccess: { ...NOT_A_KIBANA_USER, description: 'user with no access' },
+    superuser: { ...SUPERUSER, description: 'superuser' },
+    legacyAll: { ...KIBANA_LEGACY_USER, description: 'legacy user' },
+    allGlobally: { ...KIBANA_RBAC_USER, description: 'rbac user with all globally' },
+    readGlobally: {
+      ...KIBANA_RBAC_DASHBOARD_ONLY_USER,
+      description: 'rbac user with read globally',
+    },
+    dualAll: { ...KIBANA_DUAL_PRIVILEGES_USER, description: 'dual-privileges user' },
+    dualRead: {
+      ...KIBANA_DUAL_PRIVILEGES_DASHBOARD_ONLY_USER,
+      description: 'dual-privileges readonly user',
+    },
+  };
+
+  interface Modifier {
+    modifier?: T;
+  }
+  interface Security extends Modifier {
```

Comment:
I had a few different iterations of this `getTestScenarios` method, and at one point I needed to define the interfaces this way to avoid duplicating a bunch of code. However, looking at it now I think I can get rid of the `Modifier` interface entirely and define these interfaces like so:

```
  interface Security {
    modifier?: T;
    users: Record<
      | keyof typeof commonUsers
      | 'allAtDefaultSpace'
      | 'readAtDefaultSpace'
      | 'allAtSpace1'
      | 'readAtSpace1',
      TestUser
    >;
  }
  interface SecurityAndSpaces {
    modifier?: T;
    users: Record<
      keyof typeof commonUsers | 'allAtSpace' | 'readAtSpace' | 'allAtOtherSpace',
      TestUser
    >;
    spaceId: string;
  }
  interface Spaces {
    modifier?: T;
    spaceId: string;
  }
```